### PR TITLE
Create project table manifests, move entities to new project

### PR DESCRIPTION
**Pull request by mialy-defelice**
The purpose of this PR is to allow users the ability to recreate all a projects manifests as a `table` and add annotations to the manifest csvs. In addition it provides a function that allows users to move all entities represented in manifests to a new project. This will help with the limits on the number of folders allowed for a single project. 

These functions were created to address a use case by HTAN where they kept hitting file limits. With materialized views we no longer need entities, but want to save previously created ones. 

Holding off on actually pushing to develop till we have moved everything from HTAN just to make sure its working exactly as we want. 

---
From @GiaJordan 
This PR introduces a new command to schematic `schematic manifest migrate` to recreate manifests as `table` type instead of `entity` type, and to  migrate folder entities to a different project for archival. 
PR at merge includes functionality to "recreate all a projects manifests as a `table` and add annotations to the manifest CSVs", and allows users to "move all entities represented in manifests to a new project."
`schematic manifest migrate` takes the commands:

- `--project_scope`: a comma separated list of synIDs for projects on synapse
- `--archive_project`: a single synID for a project on synapse to store the archived folder entities
- `--jsonld`: the path to the user `jsonld` model
- `--return_entities`: a flag to include when it is desired to restore the archived entities to their original locations
- `--dry_run`: perform a dry run without uploading manifests or moving entities, log info about what would be moved

When archiving entities, a folder hierarchy is created in the archival project to avoid encountering Synapse limits on folder number. Entities are sorted first by source project, then by dataset. Logic is also included to only move entities that are present in a manifest and of type `folder`. This logic is stable currently, but will need to be updated if support is added for annotating folders that contain many files, instead of the files themselves.

### CODE REVIEW

**Comment by GiaJordan** [comment]
from #861: 
> Updates:
> The files were moved because they were included in a manifest, and the code was structured in a way that assumed every entity within a dataset manifest was a folder entity to hold annotations; this has now been rectified.
> I've also included a command to migrate entities to a different project for archival, or to move archived entities back to their original folders in case that use case arises again in the future, along with re-uploading manifests as tables.
> Before merging, I'd like to also do some work on managing how the archival project is structured to avoid running into the folder limits synapse imposes on the main project directory.

**Comment by GiaJordan** [comment]
Fix for issue #861 is included in this PR
@adamjtaylor @aclayton555 @vthorsson

**Comment by milen-sage** [comment]
@GiaJordan when you think we're ready to pull the trigger and start moving entities, we are good to go. I'd start with one project first to see how things are going and check with @adamjtaylor. I can take a look too if you point me to it. 🤞 it will be successful and we can move entities for all projects.

File: schematic/store/synapse.py [hunk 1]
```diff
@@ -603,10 +604,91 @@ def upload_project_manifests_to_synapse(self, projectId: str) -> List[str]:
...
+
+                annotation_entities = self.storageFileviewTable[
+                        (self.storageFileviewTable['id'].isin(manifest_df['entityId']))
+                        & (self.storageFileviewTable['type'] == 'folder')
+                    ]['id']
+
+                if returnEntities:
+                    for entityId in annotation_entities:
+                        self.syn.move(entityId, datasetId)
+                        pass
```
Comment (andrewelamb) [inline schematic/store/synapse.py:672]: What is this pass statement for?

File: schematic/store/synapse.py [hunk 2]
```diff
@@ -603,10 +604,91 @@ def upload_project_manifests_to_synapse(self, projectId: str) -> List[str]:
...
+        '''
+        Purpose:
+            For all manifests in a project, upload them as a table and add annotations manifest csv.
+            Assumes the manifest is already present as a CSV in a dataset in the project.
+
+        '''
+
+        sg = SchemaGenerator(path_to_json_ld)
+        manifests = []
+        manifest_loaded = []
```
Comment (andrewelamb) [inline schematic/store/synapse.py:621]: These lists are unused currently. Are they what the comment below are referring to?

File: schematic/store/synapse.py [hunk 3]
```diff
@@ -603,10 +604,91 @@ def upload_project_manifests_to_synapse(self, projectId: str) -> List[str]:
...
+    def upload_annotated_project_manifests_to_synapse(self, projectId:str, path_to_json_ld: str) -> List[str]:
+        '''
+        Purpose:
+            For all manifests in a project, upload them as a table and add annotations manifest csv.
+            Assumes the manifest is already present as a CSV in a dataset in the project.
+
+        '''
+
+        sg = SchemaGenerator(path_to_json_ld)
+        manifests = []
```
Comment (andrewelamb) [inline schematic/store/synapse.py:620]: This list is unused.

File: schematic/store/synapse.py [hunk 4]
```diff
@@ -603,10 +604,91 @@ def upload_project_manifests_to_synapse(self, projectId: str) -> List[str]:
...
+
+            manifest_info = self.getDatasetManifest(datasetId, downloadFile=True)
+
+            if manifest_info:
+                manifest_id = manifest_info["properties"]["id"]
+                manifest_name = manifest_info["properties"]["name"]
+                manifest_path = manifest_info["path"]
+                manifest_df = load_df(manifest_path)
+                #manifest_syn_id = self.annotate_upload_manifest_table(manifest_df, datasetId, manifest_path)
+                manifest_syn_id = self.associateMetadataWithFiles(sg, manifest_path, datasetId, manifest_record_type='table')
```
Comment (andrewelamb) [inline schematic/store/synapse.py:637]: Several of the variable in lines 632-637 are not used

File: schematic/store/synapse.py [hunk 5]
```diff
@@ -603,10 +604,91 @@ def upload_project_manifests_to_synapse(self, projectId: str) -> List[str]:
...
+
+
+    def move_entities_to_new_project(self, projectId: str, newProjectId: str, returnEntities: bool = False):
+        """
+        For each manifest csv in a project, look for all the entitiy ids that are associated.
+        Look up the entitiy in the files, move the entity to new project.
+        """
+
+        manifests = []
+        manifest_loaded = []
```
Comment (andrewelamb) [inline schematic/store/synapse.py:649]: These lists aren't currently used

File: schematic/store/synapse.py (hunk 1 above)
Comment (GiaJordan) [reply schematic/store/synapse.py:672]: I'm leaving them in until after the dry runs with HTAN are complete, they just helped with debugging

File: schematic/store/synapse.py (hunk 4 above)
Comment (GiaJordan) [reply schematic/store/synapse.py:637]: The currently unused variables are portions of the code from when @/mialy-defelice was developing that I haven't wanted to remove yet

**Review by andrewelamb — APPROVED** [review]

**Comment by GiaJordan** [comment]
@adamjtaylor merged

**Comment by GiaJordan** [comment]
@aclayton555 @adamjtaylor @vthorsson
merged

**Pull request merged**